### PR TITLE
sql: add tracing span on client side of SetupFlow RPC

### PR DESCRIPTION
**Pull request by yuzefovich**
This commit restores the creation of the tracing span for the clients side of the SetupFlow RPC when it is issued in parallel (by the DistSQL worker). This was lost in a recent refactor.

Epic: None

Release note: None

### CODE REVIEW

**Comment by yuzefovich** [comment]
This still doesn't work - I believe this is due to us forking the span with the detached recording, right? Someone still needs to collect the trace, no?

**Review by andreimatei — COMMENTED** [review]
I probably said `ForkSpan` but I wanted to say `ChildSpan`

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-:-NGrfkN0-cpRg8cx5isp:b-7dm6e3)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by yuzefovich** [comment]
Indeed, that was it, thanks!

**Review by andreimatei — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-:-NGrwlqGCY23ZaOkyK4S:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGrvh47A6ns3QCbEKON:-NGrvh47A6ns3QCbEKOO:b-iegkxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9e50b5f191736269561d9208d938c7e532716ee/pkg/sql/distsql_running.go#L477)):*
> ```Go
> 	// - now the client-side goroutine of the RPC is scheduled, and it attempts
> 	//   to use the span from the context, but it has already been finished.
> 	runnerCtx, runnerSpan := tracing.ChildSpan(origCtx, "SetupFlow" /* opName */)
> ```

Don't lie by introducing `runnerCtx` yet, only to overwrite it later.
Do 
```
evalCtx.Tracer().StartSpanCtx(ctx, "SetupFlow", WithParent(tracing.SpanFromContext(ctx)))
```

The `WithParent` part is robust to `ctx` not having a span.

And you probably also want a better name for the span. Maybe "SetupFlows async"?



_Code quote:_
```Go
runnerCtx, runnerSpan := tracing.ChildSpan(origCtx, "SetupFlow" /* opName */)
```
___
*[`pkg/sql/distsql_running.go` line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGrwcz3EEWeQzg92M0t:-NGrwcz4Cmb__ghdV6sM:b-ix65xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9e50b5f191736269561d9208d938c7e532716ee/pkg/sql/distsql_running.go#L477)):*
> ```Go
> 	// - now the client-side goroutine of the RPC is scheduled, and it attempts
> 	//   to use the span from the context, but it has already been finished.
> 	runnerCtx, runnerSpan := tracing.ChildSpan(origCtx, "SetupFlow" /* opName */)
> ```

consider putting a comment around here saying that this span is necessary because it can outlive its parent.

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei)

___
*[`pkg/sql/distsql_running.go` line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGrvh47A6ns3QCbEKON:-NGs2SugDnzONPKi_TBz:bn7jxt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9e50b5f191736269561d9208d938c7e532716ee/pkg/sql/distsql_running.go#L477)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Don't lie by introducing `runnerCtx` yet, only to overwrite it later.
Do 
```
evalCtx.Tracer().StartSpanCtx(ctx, "SetupFlow", WithParent(tracing.SpanFromContext(ctx)))
```

The `WithParent` part is robust to `ctx` not having a span.

And you probably also want a better name for the span. Maybe "SetupFlows async"?
</blockquote></details>

Done.

Do you mind though expanding on what you mean by "not lying by introducing `runnerCtx`"?
___
*[`pkg/sql/distsql_running.go` line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGrwcz3EEWeQzg92M0t:-NGs2Sl3BzsriQ-0r_-8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9e50b5f191736269561d9208d938c7e532716ee/pkg/sql/distsql_running.go#L477)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider putting a comment around here saying that this span is necessary because it can outlive its parent.
</blockquote></details>

Done.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGrvh47A6ns3QCbEKON:-NGs8Pzx5oJvdg8-WOht:babcjjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9e50b5f191736269561d9208d938c7e532716ee/pkg/sql/distsql_running.go#L477)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Done.

Do you mind though expanding on what you mean by "not lying by introducing `runnerCtx`"?
</blockquote></details>

I misread your patch, I'm sorry. In my  defense, I have a small fever at the moment :P

I thought that `runnerCtx` ends up not inheriting from `origCtx`, but that's not the case.
Now that I read this again, your use of `tracing.ChildSpan` was just fine. I apologize. What you have now is also fine, with a bit more characters. It doesn't matter.
___
*[`pkg/sql/distsql_running.go` line 514 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGsNFICEz8glevZb4fa:-NGsNFICEz8glevZb4fb:b27d3u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/34cc6c2b61b95a9eaacc147a5e7317f235aa34fa/pkg/sql/distsql_running.go#L514)):*
> ```Go
> 			if err = runReq.run(); err != nil {
> 				runnerCleanup()
> 				return ctx, flow, err
> ```

I'm confused by this return. Don't we need to continue with the `SetupFlow` requests to the other nodes ?
I would like to verify that `runnerCleanup` is called when it needs to be, but this return is tripping me up.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 490 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGsP0h79SDrLiSGEGnj:-NGsP0h79SDrLiSGEGnk:bmk35xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/34cc6c2b61b95a9eaacc147a5e7317f235aa34fa/pkg/sql/distsql_running.go#L490)):*
> ```Go
> 	for nodeID, flowSpec := range flows {
> 		if nodeID == thisNodeID {
> 			// Skip this node.
> ```

consider adding to this comment to say that the local flow was handled above.
___
*[`pkg/sql/distsql_running.go` line 514 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGsNFICEz8glevZb4fa:-NGsOM430nl422oBUn7a:brpoev6) ([raw file](https://github.com/cockroachdb/cockroach/blob/34cc6c2b61b95a9eaacc147a5e7317f235aa34fa/pkg/sql/distsql_running.go#L514)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I'm confused by this return. Don't we need to continue with the `SetupFlow` requests to the other nodes ?
I would like to verify that `runnerCleanup` is called when it needs to be, but this return is tripping me up.
</blockquote></details>

nvm, this makes sense.
But the cleanup is a bit hard to follow. Would it be feasible to write results to the channel here for both success and failure (or just for failure, and have a counter of remote flows that we need to wait for), get rid of the `if !usedWorker` block, and then let the rest of the code run even if all RPCs were performed synchronously? Then the cleanup could be verified more easily.

Btw, it's not clear to me why the buffer on the results chan is `len(flows)-1`. If buffering is needed, where is the `-1` coming from exactly?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @andreimatei)

___
*[`pkg/sql/distsql_running.go` line 477 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGrvh47A6ns3QCbEKON:-NGtI4PtAOZ0c4Ffq7Jm:b9lcoqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c9e50b5f191736269561d9208d938c7e532716ee/pkg/sql/distsql_running.go#L477)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I misread your patch, I'm sorry. In my  defense, I have a small fever at the moment :P

I thought that `runnerCtx` ends up not inheriting from `origCtx`, but that's not the case.
Now that I read this again, your use of `tracing.ChildSpan` was just fine. I apologize. What you have now is also fine, with a bit more characters. It doesn't matter.
</blockquote></details>

No worries! Reverted back to `tracing.ChildSpan` since that seems prettier.

I hope you feel better soon :crossed_fingers: 
___
*[`pkg/sql/distsql_running.go` line 490 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGsP0h79SDrLiSGEGnj:-NGtIUly1z2_jLYD4Z1B:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/34cc6c2b61b95a9eaacc147a5e7317f235aa34fa/pkg/sql/distsql_running.go#L490)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider adding to this comment to say that the local flow was handled above.
</blockquote></details>

Done.
___
*[`pkg/sql/distsql_running.go` line 514 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGsNFICEz8glevZb4fa:-NGtItC6-bU10bWcg0Mh:bpk4h4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/34cc6c2b61b95a9eaacc147a5e7317f235aa34fa/pkg/sql/distsql_running.go#L514)):*
> Would it be feasible to write results to the channel here for both success and failure (or just for failure, and have a counter of remote flows that we need to wait for), get rid of the `if !usedWorker block`, and then let the rest of the code run even if all RPCs were performed synchronously? Then the cleanup could be verified more easily.

That could work at the expense of spinning up the new goroutine as well as issuing some of the RPCs somewhat needlessly. Is that worth it?

We could track the number of synchronous RPCs issued, and, before spinning up that goroutine, we would recv from the channel the corresponding number of times, and if there is an error, we'd avoid the new goroutine too. However, that seems more complicated than the current code is.

I did refactor the code slightly so that it is more clear when the responsibility of the cleanup is handed over to the new goroutine. Let me know what you think.

> Btw, it's not clear to me why the buffer on the results chan is `len(flows)-1`. If buffering is needed, where is the `-1` coming from exactly?

For the local flow we don't issue `SetupFlow` RPC - we just call the corresponding `setupFlow` function directly, and only if that is successful do we proceed with the remote flows' setup.

**Review by andreimatei — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-:-NGw3SO50dzW5Qxwojr3:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[`pkg/sql/distsql_running.go` line 514 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/91861#-NGsNFICEz8glevZb4fa:-NGw3KqEEksRnwitUCZP:brq0mfk) ([raw file](https://github.com/cockroachdb/cockroach/blob/34cc6c2b61b95a9eaacc147a5e7317f235aa34fa/pkg/sql/distsql_running.go#L514)):*
> I did refactor the code slightly so that it is more clear when the responsibility of the cleanup is handed over to the new goroutine. Let me know what you think.

What you have now is better than before, I think. I'm happy.

**Comment by yuzefovich** [comment]
TFTR!

bors r+
